File: src_local/agents/Support/markdown_structure_agent.ts
import { MarkdownUtil } from "../../utils/markdown_util.ts";

/**
 * Agente especialista em decomposição e normalização de estruturas Markdown.
 */
export class MarkdownStructureAgent {
    private raw: string[];
    private res: string[] = [];
    private state = { seen: {} as Record<string, number>, inCb: false };

    constructor(rawLines: string[]) {
        this.raw = rawLines;
    }

    /**
     * Executa a limpeza estrutural pesada (Refinement).
     */
    executeRefinement(): string[] {
        for (let i = 0; i < this.raw.length; i++) {
            const line = this.raw[i];
            if (line === undefined) continue;
            if (this.processLine(line, i)) {
                continue;
            }
        }
        return this.res;
    }

    private processLine(line: string, idx: number): boolean {
        const stripped = line.trimEnd();

        // Gerenciamento de blocos de código
        if (this.handleCodeBlock(stripped, line)) {
            return true;
        }

        // Processamento de Cabeçalhos
        if (stripped.trim().startsWith("#")) {
            this.applyHeaderLogic(stripped.trim(), idx);
            return true;
        }

        // Conteúdo regular
        if (stripped.trim() !== "" || (this.res.length > 0 && this.res[this.res.length - 1] !== "")) {
            this.res.push(stripped);
        }
        return false;
    }

    private handleCodeBlock(stripped: string, original: string): boolean {
        if (stripped.trim().startsWith("```")) {
            this.state.inCb = !this.state.inCb;
            this.res.push(original);
            return true;
        }
        if (this.state.inCb) {
            this.res.push(original);
            return true;
        }
        return false;
    }

    private applyHeaderLogic(line: string, idx: number): void {
        if (this.res.length > 0 && this.res[this.res.length - 1] !== "") {
            this.res.push("");
        }

        const h = MarkdownUtil.deduplicateHeader(line, this.state.seen);
        this.res.push(h);
        MarkdownUtil.applyHeaderPadding(this.res, this.raw, idx);
    }
}

File: legacy_restore/src_local/agents/Support/markdown_structure_agent.py
import re
from src_local.utils.markdown_util import MarkdownUtil

class MarkdownStructureAgent:
    """Agente especialista em decomposição e normalização de estruturas Markdown."""
    
    def __init__(self, raw_lines):
        self.raw = raw_lines
        self.res = []
        self.state = {"seen": {}, "in_cb": False}

    def execute_refinement(self) -> list:
        """Executa a limpeza estrutural pesada."""
        for i, line in enumerate(self.raw):
            if self._process_line(line, i):
                continue
        return self.res

    def _process_line(self, line, idx):
        stripped = line.rstrip()
        
        # Gerenciamento de blocos de código
        if self._handle_code_block(stripped, line):
            return True
            
        # Processamento de Cabeçalhos
        if stripped.strip().startswith('#'):
            self._apply_header_logic(stripped.strip(), idx)
            return True
            
        # Conteúdo regular
        if stripped.strip() or (self.res and self.res[-1]):
            self.res.append(stripped)
        return False

    def _handle_code_block(self, stripped, original):
        if stripped.strip().startswith('```'):
            self.state["in_cb"] = not self.state["in_cb"]
            self.res.append(original)
            return True
        if self.state["in_cb"]:
            self.res.append(original)
            return True
        return False

    def _apply_header_logic(self, line, idx):
        if self.res and self.res[-1]: 
            self.res.append('')
        
        h = MarkdownUtil.deduplicate_header(line, self.state["seen"])
        self.res.append(h)
        MarkdownUtil.apply_header_padding(self.res, self.raw, idx)
